### DISCUSSION TITLE
About downloading a remote config

### DISCUSSION THREAD

**Discussion by lewis-yeung** [JanDeDobbeleer/oh-my-posh#2167] [Ideas]
I'm testing logic about how OMP uses a remote config and I gets two ideas on it. Ready to create a new PR.

### 💡 ~~Idea 1: Support both HTTPS & HTTP URLs~~

~~A remote config URL may be an HTTP one, e.g., from a private site instead of `raw.githubusercontent.com` only.~~

### 💡 Idea 2: Do not resolve the URL when a download fails

When I test `oh-my-posh init pwsh` in a bare PowerShell (launched with `-NoProfile` option) on Windows and **without an Internet connection**, it shows:

```
(@(&"<path_to_omp>/oh-my-posh.exe" init pwsh --config="<current_working_dir>\https:\raw.githubusercontent.com\JanDeDobbeleer\oh-my-posh\v7.71.2\themes\default.omp.json" --print) -join "`n") | Invoke-Expression
```

That means when the download of a remote config fails, the URL will be mistakenly resolved to an absolute path by joining with the CWD.

https://github.com/JanDeDobbeleer/oh-my-posh/blob/623fabcef31853255fa39afcf43cf7138d69079a/src/environment/shell.go#L236-L240

https://github.com/JanDeDobbeleer/oh-my-posh/blob/623fabcef31853255fa39afcf43cf7138d69079a/src/environment/shell.go#L251-L255

In most cases it will not cause a problem since the fake path will not actually flow into `$env:POSH_THEME` in PowerShell. However, when I test `oh-my-posh init zsh` in zsh on WSL, it shows:

```
export POSH_THEME="<current_working_dir>/https:/raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/v7.71.2/themes/default.omp.json"
export POWERLINE_COMMAND="oh-my-posh"
export CONDA_PROMPT_MODIFIER=false
...
```

The env `POSH_THEME` represents an invalid path in this case.

I think we can make the check like this:

```go
if err := env.downloadConfig(env.CmdFlags.Config); err != nil {
	// make it use default config when download fails
	env.CmdFlags.Config = ""
	return
}
```

where an early return is reasonable since a default config will definitely be used in this case.

</details>

**Comment by JanDeDobbeleer** [comment]
Idea 1 is not an option as that's a potential security issue. I deliberately only allow https. 

Everything else is good as that was also still on my list as I noticed it somewhere last week during another fix.

**Reply by lewis-yeung** [reply]
🆗 Got it. BTW, a space should be added in this template field of the default config:

https://github.com/JanDeDobbeleer/oh-my-posh/blob/4d048b265808ffc5cf718098f75b79f414e48b24/src/engine/config.go#L284

as `Template: " \uE23A ",`.

And in this line:

https://github.com/JanDeDobbeleer/oh-my-posh/blob/4d048b265808ffc5cf718098f75b79f414e48b24/src/environment/shell.go#L243

should `"constants.BASH"` be changed to `"bash"`?

**Reply by JanDeDobbeleer** [reply]
BASH resolves to that string.

**Reply by lewis-yeung** [reply]
Sorry but I don't quite understand how can this happen... Isn't it a string literal surrounded by double quotes? I can only  find a constant definition `BASH  = "bash"` in package `shell`:

https://github.com/JanDeDobbeleer/oh-my-posh/blob/0472cad2579ea855b7f5fa2139b873182b81ddfb/src/shell/constants.go#L5

Besides, in Git Bash I tried something like `eval "$(oh-my-posh init bash --config "/c/path/to/my_theme.omp.json")"`, and then `echo $POSH_THEME`, which showed a Windows-style path (with backslashes). This didn't seem to be the expected result. (I have read #1541.)

**Reply by JanDeDobbeleer** [reply]
You're right, I missed the quotes. They need to be removed indeed. Using `shell.BASH` is what we need.

**Reply by lewis-yeung** [reply]
Using `shell.BASH` leads to an import cycle between package `environment` and `shell`, should we use `"bash"` directly?

**Reply by JanDeDobbeleer** [reply]
Yes, no other choice in that case.

**Reply by lewis-yeung** [reply]
@JanDeDobbeleer Will these be fixed in the next release? 😄
